### PR TITLE
Fix #977, #979: improve performance of My Feeds page

### LINKED ISSUE #977: Improve performance of My Feeds page
I'm not sure what's happening, but when I try to enter a new URL for a blog feed in /myfeeds, the browser lags, and eventually shows a popup with other URLs (not sure what they are), and errors about my URL not being a feed.

This isn't a great experience.  Can we make this faster?  Are we trying to validate the feed URL?  If so, can we do that when you submit it, and let the user fix if it's wrong?

Screen Shot 2020-04-09 at 10 44 23 PM

### LINKED ISSUE #979: Can't add/delete multiple feeds one after another in /myfeeds
I add a feed to /myfeeds, and it works.  However, I'm not able to then add another.  If I click the + sign, it doesn't show another empty area with space to enter URL and Name, I get a 409 error:

Screen Shot 2020-04-09 at 10 51 40 PM

409 Conflict means something to a web developer, but it won't to a user.  Can we do a better job of letting them know what this error means, and also provide a new empty set of boxes for entering another feed?

### PR DESCRIPTION
**Pull request by Silvyre**
## Issue This PR Addresses

Fix #977
Fix #979

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

- As per #977, adding a feed to the My Feeds list was being slowed down by frequent re-rendering of the large ExistingFeedList component. This PR resolves this issue by [memoizing the ExistingFeedList component](https://github.com/Seneca-CDOT/telescope/compare/master...Silvyre:issue-977?expand=1#diff-2117800b114a1b4cfaa8fb1e262360eaR89).
- Minor fixes include:
   - Moving some logic from the MyFeeds component into the child ExistingFeedList component
   - Delaying form field validation until submission
   - Replacing TextValidator elements within ExistingFeedList with simpler TextFields elements
   - Clearing the blog feed URL form field upon submitting a new feed
   - Various other changes mentioned in https://github.com/Seneca-CDOT/telescope/issues/979#issuecomment-613280576
- When a feed is added/deleted, it now appears/disappears within the ExistingFeedList without needing to refresh the page

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx [hunk 1]
```diff
@@ -1,11 +1,39 @@
...
+import useSiteMetadata from '../../hooks/use-site-metadata';
 import DeleteFeedDialogButton from './DeleteFeedDialogButton.jsx';
 
-function ExistingFeedList({ feedHash }) {
+function ExistingFeedList({ userInfo }) {
+  const [feedHash, updateFeedHash] = useState({}); // { id1: {author: '...', url: '...'}, id2: {...}, ... }
+  const { telescopeUrl } = useSiteMetadata();
+
+  useEffect(() => {
+    if (userInfo.id) {
```
Comment (humphd) [inline src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx:13]: Invert this logic so you don't have a huge indent on all the code.  Return early if there's no point being here.

```js
if(!userInfo.id) {
  return;
}
...
```

File: src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
+          const userFeeds = await response.json();
+          const userFeedHash = userFeeds.reduce((hash, feed) => {
+            hash[feed.id] = { author: feed.author, url: feed.url };
+            return hash;
+          }, {});
+
+          updateFeedHash(userFeedHash);
+        } catch (error) {
+          console.log('Error hashing user feeds', error);
+          throw error;
```
Comment (humphd) [inline src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx:31]: No point throwing this, as there's nothing to catch it.

File: src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx (hunk 1 above)
Comment (Silvyre) [reply src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx:13]: Fixed in https://github.com/Seneca-CDOT/telescope/pull/994/commits/8a099c8ad1df285770020ddd3860102cd9ecc057

File: src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx (hunk 2 above)
Comment (Silvyre) [reply src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx]: Removed `throw` call in https://github.com/Seneca-CDOT/telescope/pull/994/commits/8a099c8ad1df285770020ddd3860102cd9ecc057

**Comment by Silvyre** [comment]
Some unexpected behaviour has led me to revert some changes within https://github.com/Seneca-CDOT/telescope/pull/994/commits/4814cfae790c647655f46256b72374d1598d4be0, namely

>- Moving some logic from the MyFeeds component into the child ExistingFeedList component

I am getting closer to fixing the regression mentioned in PR's description:

>- Fix a regression: when a feed is added, it no longer appears within the ExistingFeedList until the page is refreshed

To this end, I am experimenting with passing a `numFeeds` prop to the ExistingFeedList. My current issue is, despite this prop evidently incrementing after a the new feed is added, that feed is still not being rendered to the page.

**Comment by Silvyre** [comment]
> My current issue is, despite this prop evidently incrementing after a the new feed is added, that feed is still not being rendered to the page.

Since I believe we care a little bit more about this PR's title than fixing this regress, I'm going to open this PR to review. If we decide to land this PR, I'll open an issue to fix the regression.

**Comment by cindyledev** [comment]
I've tested this and it definitely runs a lot faster. I'm not confident in my ability to actually review this so if someone could give a first pass, I can re-review.

File: src/frontend/src/pages/myfeeds.js [hunk 3]
```diff
@@ -90,24 +91,31 @@ export default function MyFeeds() {
...
-            return hash;
-          }, {});
-
-          updateFeedHash(userFeedHash);
-        } catch (error) {
-          console.log('Error hashing user feeds', error);
-          throw error;
-        }
-      })();
+    console.log('feedHash updated, setting numFeeds to', Object.keys(feedHash).length);
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:94]: Let's dump this

File: src/frontend/src/pages/myfeeds.js [hunk 4]
```diff
@@ -90,24 +91,31 @@ export default function MyFeeds() {
...
+
+        const userFeedHash = userFeeds.reduce((hash, feed) => {
+          hash[feed.id] = { author: feed.author, url: feed.url };
+          return hash;
+        }, {});
+
+        updateFeedHash(userFeedHash);
+      } catch (error) {
+        console.log('Error hashing user feeds', error);
+        throw error;
```
Comment (humphd) [inline src/frontend/src/pages/myfeeds.js:115]: Remove this, there's nothing to catch it.

File: src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx [hunk 5]
```diff
@@ -1,12 +1,11 @@
...
-import { TextValidator } from 'react-material-ui-form-validator';
-import { Grid, Typography } from '@material-ui/core';
+import { Grid, TextField, Typography } from '@material-ui/core';
 import { AccountCircle, RssFeed } from '@material-ui/icons';
 import DeleteFeedDialogButton from './DeleteFeedDialogButton.jsx';
 
-function ExistingFeedList({ feedHash }) {
-  if (Object.keys(feedHash).length) {
+function ExistingFeedList({ feedHash, numFeeds }) {
+  if (numFeeds) {
```
[the remark this answers is not in the document]
Comment (Silvyre) [reply src/frontend/src/components/MyFeedsPage/ExistingFeedList.jsx]: Refactored return to use ternary operator in https://github.com/Seneca-CDOT/telescope/pull/994/commits/4aa793648200abf1b74f2034793548487cf486db

File: src/frontend/src/pages/myfeeds.js (hunk 3 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: Done in https://github.com/Seneca-CDOT/telescope/pull/994/commits/4aa793648200abf1b74f2034793548487cf486db

File: src/frontend/src/pages/myfeeds.js (hunk 4 above)
Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js]: Done in https://github.com/Seneca-CDOT/telescope/pull/994/commits/4aa793648200abf1b74f2034793548487cf486db

**Comment by Silvyre** [comment]
Regression fixed in https://github.com/Seneca-CDOT/telescope/pull/994/commits/14e26d9ae2670ee32a377e7a5f232bf95b246c65. I have also removed some of the unnecessary `numFeeds` logic.

**Comment by Silvyre** [comment]
I've pushed an update that likewise removes feeds from the rendered ExistingFeedList upon their deletion.

(I understand that the same functionality was developed by @c3ho for #993, but I am doubtful that that fix would prove durable after merging with the changes this PR introduces. Sorry, @c3ho!)

**Comment by cindyledev** [comment]
I just finished reviewing #993 [here](https://github.com/Seneca-CDOT/telescope/pull/993#pullrequestreview-392564795)

and I noticed that removing a feed in #993 is faster there than here.

When I try to delete a feed here, the feed doesn't disappear immediately so my first instinct was to click on the garbage bin icon to delete again and I hit this:
Annotation 2020-04-13 225700

**Comment by Silvyre** [comment]
> When I try to delete a feed here, the feed doesn't disappear immediately so my first instinct was to click on the garbage bin icon to delete again and I hit this:
> 

@cindyledev Added some logic that should hopefully prevent this from happening again, https://github.com/Seneca-CDOT/telescope/pull/994/commits/34fe501766afc707431b7a822a3d8cef71250dee.

File: src/frontend/src/pages/myfeeds.js [hunk 6]
```diff
@@ -142,14 +115,16 @@ export default function MyFeeds() {
...
+    const updatedHash = { ...feedHash };
+    delete updatedHash[id];
+    updateFeedHash(updatedHash);
   }
 
   return userInfo.id ? (
     <PageBase title="My Feeds">
       <div className={classes.margin}>
-        <ValidatorForm onSubmit={addFeed}>
+        <ValidatorForm onSubmit={addFeed} instantValidate={false}>
```
Comment (c3ho) [inline src/frontend/src/pages/myfeeds.js:127]: I think @humphd commented on this in my current PR it should be `onSubmit={() => addFeed()}`

Comment (Silvyre) [reply src/frontend/src/pages/myfeeds.js:127]: Good to know, thanks. I saw that you already made this change in your PR. When I'm fixing the merge conflicts between our work, I'll favour yours, in this situation.

**Pull request merged**